### PR TITLE
Major refactoring of digest class, entity reference resolution is now…

### PR DESCRIPTION
**Pull request by dchaplinsky**
… done inside the make_entities helper, signature of the helper changed too, entity id keys now might include links to other entities with proper dereferencing. Also * is now supported in the entity.* syntax of keys

### CODE REVIEW

**Comment by dchaplinsky** [comment]
Hey, I am a bit braindead today, so I made changes which seems to work but not the tests for them. Please try them and add test coverage (including negative tests, where the wrong key is used or circular references are present) and also compare the performance after the change.

**Comment by Legless** [comment]
Asterisk keys fails for entities with `entity` typed properties. 

See example: 
[ukrainian_mps.yaml.txt](https://github.com/dchaplinsky/thebeast/files/10832519/ukrainian_mps.yaml.txt)

```yaml
ukrainian_mp_membership:
  schema: Membership
  keys:
    - entity.*
  properties:
    organization:
      entity: verkhovna_rada
    member:
      entity: ukrainian_mp_person
```

Error:
```
  File "/usr/src/thebeast/thebeast/digest/abstract.py", line 76, in make_entities
    key_values += context_entities_map[val]
TypeError: 'NoneType' object is not iterable
```

Or with other mapping it can be: 
```
KeyError: '1832619ac2785e89db024e7687208d483472df86'
```

**Comment by dchaplinsky** [comment]
That's the exception you've asked in the case when something cannot be dereferenced yet. Make sure that the order of the entities is correct. Try to double check it by reproducing list of fields manually to see what fails.

**Comment by Legless** [comment]
Did some performance tests.

#### For asterisk keys:
Ingesting with 16 threads with single entity in mapping.

with asterisk keys: `27:20`, 2730.42it/s
no asterisk keys: `25:27`, 2932.45it/s
same mapping on `main` branch: `25:27`, 2932.45it/s

So it's a slightly slower than setting keys by hand, but I guess it's still fine. 

#### For entity keys:

Ingesting with 16 threads with 3 entities in mapping, one using `entity` keys.

this PR, with entity keys: `34:24, 2169.01it/s`
on `main` branch: `34:34`, 2159.82it/s

**Comment by Legless** [comment]
> That's the exception you've asked in the case when something cannot be dereferenced yet. Make sure that the order of the entities is correct. Try to double check it by reproducing list of fields manually to see what fails.

Ok, found it: it fails when any referenced entity is null / not generated because of empty key. 

Take a look here:
```yaml
entities:
  person:
    schema: Person
    keys:
      - entity.name
    properties:
      name:
        literal: "Ivan Sraka"

  company:
    schema: Company
    keys:
      - entity.name
    properties:
      name:
        literal: ""

  employ:
    schema: Employment
    keys:
      - entity.employer
      - entity.employee
    properties:
      employer: 
        entity: company
      employee: 
        entity: person
```

`company` will not be generated due to empty key, and `employ` will fail with exception:
```
TypeError: 'NoneType' object is not iterable
```

Maybe then it should not fail, but gracefully return empty string if entity can't be resolved, as with simple `record.foobar`-style key? Because this behavior renders this feature kinda unusable on real data, e.g.: 

| person_name      | work_place    |
|------------------|---------------|
| Ivan Sraka       | TOV Romashka  |
| Petro Poroshenko | */empty value/* |

**Comment by Legless** [comment]
Added the test, but just saw I've forgoted this:
> including negative tests, where the wrong key is used or circular references are present

Sorry, will do this also.

**Comment by Legless** [comment]
I've  added tests for wrong entity order, but couldnt think of any circular reference mapping example.

File: thebeast/tests/test_entity_keys.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+        entities = list(ent.payload for ent in mapping.digestor.extract(items))
+        self.assertEqual(6, len(entities))
+
+        self.assertEqual("Employment", entities[0]["schema"])
+        self.assertEqual("Company", entities[1]["schema"])
+        self.assertEqual("Person", entities[2]["schema"])
+
+        # both employment ids will be `None` since it can't resolve it's related entities, being before them in mapping
+        self.assertEqual(None, entities[0]["id"])
```
Comment (dchaplinsky) [inline thebeast/tests/test_entity_keys.py:117]: Please add a test to verify that entities without ids are marked as invalid (that was a change to the logic as well)

File: thebeast/tests/test_entity_keys.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+                }
+            ),
+        ]
+
+        entities = list(ent.payload for ent in mapping.digestor.extract(items))
+        self.assertEqual(3, len(entities))
+
+        # ids must differ with each row
+        self.assertEqual("c1d4c4652b8b9a8d5cda22d508e497a12dd9d07f", entities[0]["id"])
+        self.assertEqual("c131fc0732f46173b22c8698d0cc5d0d220a7932", entities[1]["id"])
```
Comment (dchaplinsky) [inline thebeast/tests/test_entity_keys.py:38]: It's creepy, two sha1 that has the same prefix.

File: thebeast/tests/test_entity_keys.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+            ]
+
+            entities = list(ent.payload for ent in mapping.digestor.extract(items))
+            self.assertEqual(6, len(entities))
+
+            self.assertEqual("Person", entities[0]["schema"])
+            self.assertEqual("Company", entities[1]["schema"])
+            self.assertEqual("Employment", entities[2]["schema"])
+
+            self.assertEqual("17dec550a297076989f408f075db145c3f72b80a", entities[0]["id"])
```
Comment (dchaplinsky) [inline thebeast/tests/test_entity_keys.py:86]: Oh my god, oh my gooood, omfg!

File: thebeast/tests/test_entity_keys.py (hunk 1 above)
Comment (Legless) [reply thebeast/tests/test_entity_keys.py]: Done in f806bec51e95ad6b7ddfe15c5372e6e3d1fd52fc
